src/broyden.jl

```diff
@@ -0,0 +1,170 @@
+# based on 
+# A QUASI-NEWTON METHOD FOR SOLVING SMALL NONLINEAR SYSTEMS OF ALGEBRAIC EQUATIONS
+# JEFFREY MARTIN
+# http://web.mit.edu/jmartin3/Public/Project.pdf
+macro broydentrace(stepnorm)
+    quote
+        if tracing
+            dt = Dict()
+            if extended_trace
+                dt["x"] = copy(x)
+                dt["f(x)"] = copy(fvec)
+                dt["g(x)"] = copy(fjac)
+            end
+            update!(tr,
+                    it,
+                    maximum(abs(fvec)),
+                    $stepnorm,
+                    dt,
+                    store_trace,
+                    show_trace)
+        end
+    end
+end
+
+
+const τ = 0.5
+const σ1 = 1e-8
+const σ2 = 1e-8
+const ρ = 1.0 - 1e-8
+const η = 1e-8
+const restol = 1e-6
+const imax = 10
+function broyden_{T}(df::AbstractDifferentiableMultivariateFunction,
+                   initial_x::Vector{T},
+                   xtol::T,
+                   ftol::T,
+                   iterations::Integer,
+                   store_trace::Bool,
+                   show_trace::Bool,
+                   extended_trace::Bool)
+
+    x = copy(initial_x)
+    nn = length(x)
+    xold = fill(convert(T, NaN), nn)
+    fvec = Array(T, nn)
+    fvecold = similar(fvec)
+    fjac = alloc_jacobian(df, T, nn)
+    scaleinvfjac = alloc_jacobian(df, T, nn)
+    oldinvfjac = alloc_jacobian(df, T, nn)
+    minvjΔf = similar(xold)
+    p = Array(T, nn)
+    g = Array(T, nn)
+    gr = Array(T, nn)
+
+    # Count function calls
+    f_calls::Int, g_calls::Int = 0, 0
+
+    df.fg!(x, fvec, fjac)
```

Comment:
One of the main attractions of the Broyden method is for when the Jacobian is not available.  Is there a way to use a finite-difference Jacobian as a fallback here?  It doesn't need to be super-accurate for Broyden to make progress.